### PR TITLE
Load lldp and enabled/disabled status on Kytos bootstrap

### LINKED ISSUE #124: Restore of administrative state automatic on kytos bootstrap
Today, after an unexpected quit of kytos, the user needs to execute a command through the REST Endpoint `GET \ restore` to recover and restore the administrative state saved in the storehouse. But, this recovery is expected to be automatic.

Pseudocode behavior expected:
``` 
kytos_running:
  # store administrative changes (switches, interfaces, lldp and links)
  any_change_detected:
    store_changes()
```
In kytos bootstrap
``` code
kytos_bootstrap:
  if has_changes_stored:
    _restore_changes()

```
**Note**: To use this change, it is recommended to clear the `kytos.topology.status` namespace in storehouse.

Related: kytos/kytos#1127

### PR DESCRIPTION
**Pull request by cmagnobarbosa**
Today, to restore administrative changes, it is necessary to use REST Endpoint `restore`. This PR changes the `restore_network_status` method responsible for this task, making this task automatic in the kytos bootstrap.

Main changes:
- The state stored in the storehouse is loaded into NApp `execute`.
- The states are restored when the switch or link is discovered by the topology.
- Remove the REST Endpoint `\restore`.
- In the kytos bootstrap, the state saved in the storehouse is restored.

**Note**: To use this change, it is recommended to clear the` kytos.topology.status` namespace in storehouse.

Main updates:
- Update unit tests.
- Update the storehouse client to expect to recover or create a box in the storehouse.

Fix #124

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -190,16 +197,17 @@ def get_topology(self):
...
+        """Restore the network administrative status saved in storehouse."""
         try:
-            self._load_network_status()
-            self._restore_status()
+            if isinstance(obj, Switch):
+                self._restore_switches(obj.id)
+            elif isinstance(obj, Link):
+                if obj.id not in self.links_verified:
+                    self.links_verified.append(obj.id)
+                    self._restore_links(obj.id)
```
Comment (ajoaoff) [inline main.py]:
```suggestion
                    self._restore_link(obj.id)
```

File: main.py [hunk 2]
```diff
@@ -190,16 +197,17 @@ def get_topology(self):
...
-    @rest('v3/restore')
-    def restore_network_status(self):
-        """Restore the network administrative status saved in StoreHouse."""
+    def restore_network_status(self, obj):
+        """Restore the network administrative status saved in storehouse."""
         try:
-            self._load_network_status()
-            self._restore_status()
+            if isinstance(obj, Switch):
+                self._restore_switches(obj.id)
```
Comment (ajoaoff) [inline main.py]:
```suggestion
                self._restore_switch(obj.id)
```

File: main.py [hunk 3]
```diff
@@ -91,72 +96,75 @@ def _get_link_from_interface(self, interface):
...
+                self.links[link_id].enable()
+            else:
+                self.links[link_id].disable()
+        except KeyError:
+            error = ('Error restoring link status.'
+                     f'The link {link_id} does not exist.')
+            raise KeyError(error)
+        log.info(f'The state of link {link_id} has been restored.')
+
+    def _restore_switches(self, switch_id):
```
Comment (ajoaoff) [inline main.py]: Same here.
```suggestion
    def _restore_switch(self, switch_id):
```

File: main.py [hunk 4]
```diff
@@ -91,72 +96,75 @@ def _get_link_from_interface(self, interface):
...
-
-            link = self._get_link_or_create(endpoint_a, endpoint_b)
-            endpoint_a.update_link(link)
-            endpoint_b.update_link(link)
-
-            endpoint_a.nni = True
-            endpoint_b.nni = True
-
-            self.notify_topology_update()
+    def _restore_links(self, link_id):
```
Comment (ajoaoff) [inline main.py]: As this method now restores just one link.
```suggestion
    def _restore_link(self, link_id):
```

**Review by ajoaoff — CHANGES_REQUESTED** [review]

File: tests/unit/test_main.py [hunk 5]
```diff
@@ -241,65 +223,111 @@ def test_restore_network_status(self, mock_storehouse):
...
+        mock_switch_b = get_switch_mock(dpid_b)
+        mock_interface_a_1 = get_interface_mock('s1-eth1', 1, mock_switch_a)
+        mock_interface_b_1 = get_interface_mock('s2-eth1', 1, mock_switch_b)
+        mock_link = get_link_mock(mock_interface_a_1, mock_interface_b_1)
+        mock_link.id = link_id
+        self.napp.links = {link_id: mock_link}
+        self.napp.links_state = {link_id: {"enabled": True}}
+        with self.assertRaises(KeyError):
+            self.napp._restore_link(link_id_fail)
+        self.assertRaises(KeyError)
```
Comment (gleybersonandrade) [inline tests/unit/test_main.py]: I think this 'assertRaises' can be removed, because it was tested before.

File: tests/unit/test_main.py [hunk 6]
```diff
@@ -241,65 +223,111 @@ def test_restore_network_status(self, mock_storehouse):
...
+
+        self.napp.links_state = {link_id_fail: {"enabled": True}}
+        with self.assertRaises(KeyError):
+            self.napp._restore_link(link_id_fail)
+        self.assertRaises(KeyError)
```
Comment (gleybersonandrade) [inline tests/unit/test_main.py]: Also here

File: tests/unit/test_main.py [hunk 7]
```diff
@@ -241,65 +223,111 @@ def test_restore_network_status(self, mock_storehouse):
...
+        mock_interface = get_interface_mock('s1-eth1', 1, mock_switch)
+        mock_switch.interfaces = {1: mock_interface}
+        self.napp.controller.switche = {dpid: mock_switch}
+        self.napp.switches_state = {dpid: True}
+        self.napp.interfaces_state = {'00:00:00:00:00:00:00:01:1': (True,
+                                                                    True)}
+
+        with self.assertRaises(KeyError):
+            self.napp._restore_switch(dpid_fail)
+        self.assertRaises(KeyError)
```
Comment (gleybersonandrade) [inline tests/unit/test_main.py]: Here

File: tests/unit/test_main.py [hunk 8]
```diff
@@ -241,65 +223,111 @@ def test_restore_network_status(self, mock_storehouse):
...
+
+        self.napp.switches_state = {dpid_fail: True}
+        with self.assertRaises(KeyError):
+            self.napp._restore_switch(dpid_fail)
+        self.assertRaises(KeyError)
```
Comment (gleybersonandrade) [inline tests/unit/test_main.py]: And here

**Review by ajoaoff — APPROVED** [review]

**Review by gleybersonandrade — APPROVED** [review]

File: main.py [hunk 9]
```diff
@@ -92,71 +97,79 @@ def _get_link_from_interface(self, interface):
...
+                switch.enable()
+            else:
+                switch.disable()
+        except KeyError:
+            error = ('Error while restoring switches status. The '
+                     f'{switch_id} does not exist.')
+            raise KeyError(error)
+        # wait to restore interfaces
+        while not switch.interfaces:
+            time.sleep(self.interface_restore)
```
Comment (hdiogenes) [inline main.py:152]: I think we should have a log.debug here, so we can detect if this code gets stuck for some reason.

File: settings.py [hunk 10]
```diff
@@ -5,3 +5,7 @@
 
 # Time (in seconds) to wait before setting a link as up
 LINK_UP_TIMER = 10
+# Time (in seconds) to wait retrieve box from storehouse
+WAIT_PERSISTENCE_BOX_TIMER = 0.1
```
Comment (hdiogenes) [inline settings.py]:
```suggestion
BOX_RESTORE_TIMER = 0.1
```

File: storehouse.py [hunk 11]
```diff
@@ -20,15 +25,19 @@ def __init__(self, controller):
...
+        self.list_stored_boxes()
 
     def get_data(self):
         """Return the persistence box data."""
-        if not self.box:
-            return {}
-        self.get_stored_box(self.box.box_id)
+        # Wait retrieve or create box in storehouse
+        while not self.box:
+            time.sleep(self.wait_persistence_box)
```
Comment (hdiogenes) [inline storehouse.py]: The default here is 0.1 seconds, so a debug may flood the console. Maybe a better option is to add an upper limit: it runs for X times without result, give up and log and error.

**Review by hdiogenes — CHANGES_REQUESTED** [review]

File: settings.py (hunk 10 above)
Comment (cmagnobarbosa) [reply settings.py]: Resolved.

File: storehouse.py (hunk 11 above)
Comment (cmagnobarbosa) [reply storehouse.py]: Resolved.

**Pull request merged**